### ISSUE TITLE
error on login if it reaches a standby node

### ISSUE THREAD

**Issue by Justin-DynamicD** [Caiyeon/goldfish#206]
I have a vault configuration that does _not_ use a LB in front, rather a simple rrdns to all available nodes.  Using the vault_api_addr each node will also report it's real address, and I believe allow them to redirect to the appropriate active node in the event ca client resolves a standby.

On the  vault client this functions perfectly:  I can "vault auth" and always get an appropriate response.

On goldsifh, however, selecting the "login" link I will randomly get the following warning:

```
Error making API request. URL: GET https://<generic.url>:8200/v1/sys/health Code: 429. Errors:
```

This implies I _have_ to point at the active vault, which doesn't seem proper?  Or is some way to tell goldfish it needs to be ha aware?

**Comment by Caiyeon** [comment]
I don't fully understand the workflow description.
1. Is <generic.url> the vault url or goldfish url?
2. Were you able to successfully bootstrap goldfish in the first place?

Generally speaking, HA-awareness is vault side, and based entirely in vault's API. There are known bugs in previous vault versions regarding HA, but this seems to be a different issue. https://github.com/Caiyeon/goldfish/issues/134

**Comment by Justin-DynamicD** [comment]
Sorry about that,  I was trying to mask urls to protect the innocent, blah blah.

Better detail:
I have a vault cluster with 3 nodes.  The url chosen is "vault.domain.com".  They share a single UCC/SAN certificate that contains both the chosen URL as well as their actual hostnames.  When configuring the cluster, I used the "vault_api_addr" to ensure the hostname was returned during redirects.  So, were I to query the vault, I only have a 1:3 chance of landing on the active node, but because the api_addr was set the correct hostname is known and can be returned to whoever queries it.

At least I think that's how it works (honestly I find the documentation to be rather poor).  At any rate, using the nomad agent to auth and request creds works 100% so I assume all the redirects are doing what they should.

I'm deploying goldfish with Nomad (yippee), and I don't think I have any errors but for kicks here's my job file 99% intact:

```
job "lv-ops-goldfish" {
  region = "global"
  datacenters = ["lv-ops"]
  type = "service"
  update {
    stagger      = "30s"
    max_parallel = 1
  }
  group "goldfish" {
    count = 1
    restart {
      attempts = 10
      interval = "5m"
      delay    = "25s"
      mode     = "delay"
    }
    task "goldfish-container" {
      artifact {
        source      = "git::https://<internal github>/goldfish.git//goldfish"
        destination = "apps/goldfish"
      }
      driver = "docker"
      config {
        image              = "elijahwright/goldfish"
        command            = "/app/goldfish" 
        args = [
          "-config=/app/config/docker.hcl",
          "-nomad-token-file=${NOMAD_SECRETS_DIR}/vault_token",
        ]
        port_map {
          http = 8000
        }
        volumes = [
          "apps/goldfish:/app/config",
        ]
      }
      vault {
        policies = ["goldfish"]
      }
      service {
	      name = "vault-ui"
        port = "http"
        tags = ["proxy"]
        check {
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
      resources {
        cpu    = 100
        memory = 300 # in MB
        network {
          mbits = 100
          port "http" {}
        }
      }
    }
  }
}
```
You can see I am using your nomad flag to present the secret, and I don't see an error in the logs so I assume I did it right?

My problem is  goldfish logs into vault using the line "address = https://lv-vault.domain.com:8200", but occasionally fails wit hthe error I listed above.  The error code clearly shows that it hit a standby node, which means I'm seeing one of two things:

1.  goldfish isn't handling hitting a standby correctly (either cause it has a bug or maybe my token isn't working right in Nomad?  I don't think it's the later at least ...)
2.  vault isn't redirecting goldfish correctly, although it IS redirecting the vault client/agent correctly (possible, but it means sleuthing around a bit more for said missing setting?)

**Comment by Caiyeon** [comment]
If goldfish occasionally works, then there is something wrong with the HA-setup (with your round robin load balancer). The network flow is: Goldfish -> LB -> (1 of 3) vault instances. In this scenario, Goldfish can't even tell the difference between these vault instances.
My theory is that your standby vault nodes are in redirect mode. It should be noted that redirect mode is incompatible with round robin load balancers (for obvious reasons)

**Comment by Justin-DynamicD** [comment]
There is no lb.  I'm using rrdns.  Basically 3 A records, one resolving to each vault server.  That's why goldfish will randomly hit passive nodes (again, with consul-template or a vault agent this is fine, vault merely redirects to the active node).

Sent from my mobile phone, sorry for any grammatical errors

________________________________

**Comment by Caiyeon** [comment]
Which HA-mode is vault set as?

**Comment by Justin-DynamicD** [comment]
Again, I'll blame "poor documentation", as honestly I'm not explicitly sure.  If I just look at the logs I do see a redirect address from the configuration output at start:

```
==> Vault server configuration:

                     Cgo: disabled
         Cluster Address: https://server1.domiain.com:8201
              Listener 1: tcp (addr: "0.0.0.0:8200", cluster address: "0.0.0.0:8201", tls: "enabled")
               Log Level: info
                   Mlock: supported: true, enabled: true
        Redirect Address: https://server1.domiain.com:8200
                 Storage: consul (HA available)
                 Version: Vault v0.9.0
             Version Sha: bdac1854478538052ba5b7ec9a9ec688d35a3335

==> Vault server started! Log data will stream in below:
```

But at the same time, according to documentation I should be getting a 307 if redirect is enabled and I hit a passive node, which I don't.  I get a 429?

I have both api_addr and cluster_addr defined.  I've found by toggling the two I get different behaviors, but none of them make goldfish happy :(

Going to double check the token next

**Comment by Caiyeon** [comment]
Ah I see. The 429 error only pops up on loading the login page, correct? Every other functionality works? I think it's because of this https://www.vaultproject.io/api/system/health.html
Vault's documentation actually states that different nodes give different responses for the health check, so every 2/3 hits would result in different answers.
I'm not sure if there's anything that should be done in this case. I think I've read somewhere in vault docs that round robin is actually unsuitable for HA because of how vault is setup

**Comment by Justin-DynamicD** [comment]
Yeah that's the short of it.  I find it curious that that is your conclusion is that deep should avoided, would you mind sharing the link?  I'm trying to gather as much info as possible on the topic so any info would be appreciated.  Vault has an HA page saying it's a valid setup and the other HashiCorp tools seem to behave, so I simply assumed there was a mechanism for redirect/resolving the active node.  I'll try to do more research into topic as well and see if I can she'd more light.

Sent from my mobile phone, sorry for any grammatical errors

________________________________

**Comment by Caiyeon** [comment]
Round robin works fine if there is actually a way to directly contact each node. Although, the suggested method is service discovery via consul https://groups.google.com/forum/#!msg/vault-tool/E9wLwBUkYsM/_H0hYGXhBQA

I assume in your setup there's no way to directly contact a specific node? If so, then there is no way to fix this on Goldfish. Even if I add extra logic to resolve the leader address, the address will still be sent to the leader at a 1/3 chance.

**Comment by Justin-DynamicD** [comment]
Actually there is nothing stopping direct communication to each node, which is why I'm confused.  All nodes are completely resolvable.

As for using service discovery, that actually comes with it's own problems though I guess it's not insurmountable and would solve the major problem (as I could point at active.vault.service.consul and know with 100% certainty to hit the right box) .

The problem comes from the way consul handles registering servers when they provide an fqdn vs an IP address.  Ideally my certificate should not use IP addresses, and thus in my redirect_addr I would use the fqdn of the host.  However when this is done, this will cause consul to return CNAMES when queried for said service instead of a records.  With multiple vaults registering the vault service, a dig would then return multiple CNAMES ... which technically violates RFCs and flat out breaks some other things (like dnsmasq).

So in order to get things 100% working I'd needto replace my SSL being used with one that contains:

```
active.vault.service.consul
vault.service.consul
active.vault.service.dc1.consul (gotta make sure cross datacenter works, right?)
vault.service.dc1.consul
192.168.1.1
192.168.1.2
192.168.1.3
```
This would make goldfish completely happy (using pure SD to resolve) and should result in a proper vault server with resolution, but it "feels" dirty, if that makes sense.

**Comment by Justin-DynamicD** [comment]
I think I'm going to go ahead and close this.  While I still think there may need ot be some logic to help determine the active node url, I can at least get stable operation by following above (creating a cert that accepts all possible SD combinations).

**Comment by Caiyeon** [comment]
I think this is an unfortunate result of vault's `/health` endpoint responding with the node's health rather than the entire cluster's health. The best scenario would be that Goldfish resolves the leader address and makes a second request. While I believe this should be implemented, I don't think it'll make it in the v0.8.0 release (which is coming up very soon)
